src/extensions/idle.rs

```diff
@@ -108,39 +111,84 @@ impl<'a, T: Read + Write + 'a> Handle<'a, T> {
         self.session.readline(&mut v)?;
         if v.starts_with(b"+") {
-            self.done = false;
-            return Ok(());
+            return Ok(last_idle);
         }
 
+        self.last_idle = None;
         self.session.read_response_onto(&mut v)?;
         // We should *only* get a continuation on an error (i.e., it gives BAD or NO).
         unreachable!();
     }
 
     fn terminate(&mut self) -> Result<()> {
-        if !self.done {
-            self.done = true;
+        if let Some(_) = self.last_idle.take() {
             self.session.write_line(b"DONE")?;
             self.session.read_response().map(|_| ())
         } else {
             Ok(())
         }
     }
+}
 
-    /// Internal helper that doesn't consume self.
+impl<'a, T: SetReadTimeout + Read + Write + 'a> Handle<'a, T> {
+    /// Set the timeout duration on the connection. This will also set the frequency
+    /// at which the connection is refreshed.
     ///
-    /// This is necessary so that we can keep using the inner `Session` in `wait_while`.
-    fn wait_inner<F>(&mut self, reconnect: bool, mut callback: F) -> Result<WaitOutcome>
+    /// The interval defaults to 29 minutes as given in RFC 2177.
+    pub fn timeout(&mut self, interval: Duration) -> &mut Self {
+        self.timeout = interval;
+        self
+    }
+
+    /// Do not continuously refresh the IDLE connection in the background.
+    ///
+    /// By default, connections will periodically be refreshed in the background using the
+    /// timeout duration set by [`Handle::timeout`]. If you do not want this behaviour, call
+    /// this function and the connection will simply IDLE until `wait_while` returns or
+    /// the timeout expires.
+    pub fn keepalive(&mut self, keepalive: bool) -> &mut Self {
+        self.keepalive = keepalive;
+        self
+    }
+
+    /// Block until the given callback returns `false`, or until a response
+    /// arrives that is not explicitly handled by [`UnsolicitedResponse`].
+    pub fn wait_while<F>(&mut self, mut callback: F) -> Result<WaitOutcome>
     where
         F: FnMut(UnsolicitedResponse) -> bool,
     {
         let mut v = Vec::new();
         let result = loop {
-            match self.session.readline(&mut v) {
+            match {
+                // The server MAY consider a client inactive if it has an IDLE command
+                // running, and if such a server has an inactivity timeout it MAY log
+                // the client off implicitly at the end of its timeout period.  Because
+                // of that, clients using IDLE are advised to terminate the IDLE and
+                // re-issue it at least every 29 minutes to avoid being logged off.
+                // This still allows a client to receive immediate mailbox updates even
+                // though it need only "poll" at half hour intervals.
+                let time_since_idle = Instant::now() - self.last_idle.map(Ok).unwrap_or_else(|| self.init())?;
```

Comment:
I _think_ this means that if `init` fails, we may fail to reset the read timeout to `None` if we've already gone around the loop at least one? Because the `?` here will return from the _function_, not into the match.

Reply:
True. Also from `self.terminate()?` a couple of lines below. I'll change both to break on `Err` instead of using `?`.